File: clash-hash/src/Reference/SampleInBall.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}

-- |
-- Module      : Reference.SampleInBall
-- Description : Pure Haskell Reference Implementation of FIPS 204 SampleInBall (Algorithm 29)
-- License     : MIT
-- Standard    : NIST FIPS 204 (ML-DSA), Algorithm 29
--
-- This module implements the golden specification model for SampleInBall.
-- It generates a polynomial c in R with coefficients from {-1, 0, 1}
-- and Hamming weight tau <= 64 from a 32-byte seed rho using SHAKE256 XOF.
module Reference.SampleInBall
  ( sampleInBall,
    sampleInBallTau,
    sampleInBallBS,
  )
where

import Data.Bits (testBit)
import Data.ByteString (ByteString)
import Data.ByteString qualified as BS
import Data.Int (Int32)
import Data.Word (Word8)
import Reference.Crypton (shake256)
import Prelude

-- | Polynomial representation: List of 256 coefficients in Z_q or {-1, 0, 1}
type Poly = [Int32]

-- | ML-DSA modulus q = 8380417
modQ :: Int32
modQ = 8380417

-- | FIPS 204 Algorithm 29 SampleInBall with configurable tau (e.g., 39, 49, 60).
-- Input: 32-byte seed rho.
-- Output: List of 256 coefficients in Z_q (-1 represented as q - 1 = 8380416).
sampleInBallTau :: Int -> ByteString -> Poly
sampleInBallTau tau rho =
  let -- Step 2-4: Absorb rho into SHAKE256 XOF and squeeze initial byte stream.
      -- 256 bytes is guaranteed to cover max requirement (< 221 bytes by FIPS 204 bound).
      stream = BS.unpack (shake256 256 rho)
      (signBytes, byteStream) = splitAt 8 stream

      -- Step 5: Convert 8 sign bytes into 64 bits h[0..63] (little-endian per byte)
      signBits = [testBit (signBytes !! (k `div` 8)) (k `mod` 8) | k <- [0 .. 63]]

      -- Step 1: Initialize polynomial c to 256 zeros
      initPoly = replicate 256 (0 :: Int32)

      -- Step 6-13: Fisher-Yates shuffle loop for i from (256 - tau) to 255
      go [] _ _ c = c
      go (i : is) bs hIdx c =
        -- Step 7-10: Rejection sampling to extract pseudorandom byte j <= i
        let (j, bs') = findValidByte bs i
            -- Step 12: Determine sign (-1)^h[i + tau - 256]
            signBit = signBits !! hIdx
            signVal = if signBit then modQ - 1 else 1 -- -1 in Z_q is q - 1

            -- Step 11-12: Swap c[i] <- c[j] and set c[j] <- signVal
            cj = c !! j
            c' = updatePoly i j cj signVal c
         in go is bs' (hIdx + 1) c'

      indices = [(256 - tau) .. 255]
   in go indices byteStream 0 initPoly

-- | Helper to find the first byte j <= i in the squeezed byte stream (Rejection Sampling).
findValidByte :: [Word8] -> Int -> (Int, [Word8])
findValidByte [] _ = error "SampleInBall: SHAKE256 squeeze stream unexpectedly exhausted"
findValidByte (b : bs) i =
  let j = fromIntegral b
   in if j <= i
        then (j, bs)
        else findValidByte bs i

-- | Update polynomial coefficients according to Algorithm 29 Lines 11-12.
-- If j < i: c[i] <- old c[j], c[j] <- signVal.
-- If j == i: c[i] <- signVal.
updatePoly :: Int -> Int -> Int32 -> Int32 -> [Int32] -> [Int32]
updatePoly i j cj signVal c
  | j == i = setIndex i signVal c
  | otherwise = setIndex j signVal (setIndex i cj c)

-- | Set list element at specific 0-based index.
setIndex :: Int -> a -> [a] -> [a]
setIndex idx val xs =
  let (left, _ : right) = splitAt idx xs
   in left ++ (val : right)

-- | Default SampleInBall using tau = 39 (ML-DSA-44 default).
sampleInBall :: ByteString -> Poly
sampleInBall = sampleInBallTau 39

-- | ByteString variant returning 256 coefficients as raw bytes for test harness.
sampleInBallBS :: Int -> ByteString -> ByteString
sampleInBallBS tau rho =
  let coeffs = sampleInBallTau tau rho
   in BS.pack (map fromIntegral coeffs)

File: clash-hash/reference/sample_in_ball.py
import hashlib

MOD_Q = 8380417

def sample_in_ball(rho: bytes, tau: int = 39):
    # Step 1: c = [0] * 256
    c = [0] * 256
    
    # Step 2-4: SHAKE256(rho)
    # We squeeze 8 bytes for signs, plus enough bytes for rejection sampling (e.g. 512 bytes total)
    ctx = hashlib.shake_256(rho)
    stream = ctx.digest(512)
    
    sign_bytes = stream[:8]
    byte_stream = list(stream[8:])
    
    # h bits: 64 bits from 8 bytes, little endian per byte
    h = []
    for byte in sign_bytes:
        for bit_idx in range(8):
            h.append((byte >> bit_idx) & 1)
            
    stream_idx = 0
    h_idx = 0
    for i in range(256 - tau, 256):
        # Step 6-9: Rejection sampling
        while True:
            if stream_idx >= len(byte_stream):
                raise RuntimeError("Byte stream exhausted")
            j = byte_stream[stream_idx]
            stream_idx += 1
            if j <= i:
                break
        
        # Step 10-11: Fisher-Yates update
        sign_bit = h[h_idx]
        h_idx += 1
        sign_val = MOD_Q - 1 if sign_bit == 1 else 1  # -1 or +1 in Z_q
        
        c[i] = c[j]
        c[j] = sign_val
        
    return c

if __name__ == "__main__":
    # Test with a known seed
    seed = b"\x00" * 32
    poly44 = sample_in_ball(seed, tau=39)
    non_zeros = [(idx, val) for idx, val in enumerate(poly44) if val != 0]
    print(f"ML-DSA-44 (tau=39) non-zero count: {len(non_zeros)}")
    print(f"Sample non-zero coeffs: {non_zeros[:5]}")
    
    poly65 = sample_in_ball(seed, tau=49)
    print(f"ML-DSA-65 (tau=49) non-zero count: {len([x for x in poly65 if x != 0])}")
    
    poly87 = sample_in_ball(seed, tau=60)
    print(f"ML-DSA-87 (tau=60) non-zero count: {len([x for x in poly87 if x != 0])}")
